### ISSUE TITLE
Error: NotEnoughSpace

### ISSUE THREAD

**Issue by Lusitaniae** [tinysearch/tinysearch#111]
Error
```
./tinysearch public/index.json
Error: NotEnoughSpace
```

Index file
```
 ls -lha public/index.json
-rw-r--r-- 1 q q 6,7M Jun 11 22:46 public/index.json
```

strace results
```
strace ./tinysearch public/index.json -p /tmp/

...

getrandom("", 0, GRND_NONBLOCK)         = 0
getrandom("\x1d\xee\x9e\xa5\x0c\x2d\xf4\x99\x9c\x9a\xe5\xb8\xc4\x52\x05\xb7\x4c\x3c\x2a\x79\xe4\xee\x5f\x24\xaf\x1e\xcf\x62\xb8\x49\x35\x97", 32, GRND_NONBLOCK) = 32
write(2, "Error: ", 7Error: )                  = 7
write(2, "NotEnoughSpace", 14NotEnoughSpace)          = 14
write(2, "\n", 1
)                       = 1
sigaltstack({ss_sp=NULL, ss_flags=SS_DISABLE, ss_size=8192}, NULL) = 0
munmap(0x7f7fffdd5000, 12288)           = 0
exit_group(1)                           = ?
+++ exited with 1 +++
```

**Comment by mre** [comment]
Thanks for the report. I can see that your `index.json` is quite large (6.5m). I've never experimented with such a big index.
Can you try playing with the capacity in the Cuckoofilter here?
https://github.com/mre/tinysearch/blob/6764fbec789b236abf1fba6ca8581cb8e3abcbe8/bin/src/storage.rs#L68
I'm curious how big the magic constant (that we should eventually remove) has to be to fit your index in.

**Comment by Lusitaniae** [comment]
ah, I was wondering I was getting that error, both in my local and the CI building.

I wanted to see how it was gonna work with such big index (9k items), but I was probably gonna remove body content anyway.

Will test and get back to you

**Comment by fluential** [comment]
I am actually just experimenting with same issue, thanks for the heads up

**Comment by mre** [comment]
>  UPDATE: For a file size 4535781 magic number was 56
Can you tell me how big the `wasm` file was in your case?

**Comment by fluential** [comment]
@mre 279014 bytes

**Comment by mre** [comment]
Okay thanks for the info. That sounds quite reasonable to me. Could you check the gzipped version as well if you don't mind? 😅

**Comment by fluential** [comment]
@mre you mean ?

```
gzip -9 tinysearch_engine_bg.wasm && ls -la *.gz
203966 Jun 12 15:40 tinysearch_engine_bg.wasm.gz
```

**Comment by mre** [comment]
Precisely @fluential. Thanks!

**Comment by fluential** [comment]
@mre Actually I've checked and for nginx to serve gzipped wasm requires extra config for mime types
```echo 'application/wasm                      wasm;' >> /etc/nginx/mime.types```
and adding to ```gzip_types```

via https://rustwasm.github.io/book/reference/deploying-to-production.html

**Comment by kvz** [comment]
This is happening for me too when I try to run:

```bash
$ ~/.cargo/bin/tinysearch _site/tinysearch.json
Error: NotEnoughSpace

$ du -hs _site/tinysearch.json
820K	_site/tinysearch.json
```

I just freshly installed TinySearch after reading about it on HN. Love the idea :) But I'm just kicking the tires and not in the stage of serving out wasm yet or anything so seems unrelated to what was discussed above maybe?

**Comment by kvz** [comment]
I was planning on going a little bigger than 820K, would you say TinySearch isn't the right tool for this type of job?

**Comment by mre** [comment]
I think you should try. Can you tweak the constant as described above to see what you have to set it to until it works? I'll try to fix that proper but it would help me a lot to get some more data points. :)
And thanks for testing out tinysearch in the first place.

**Comment by kvz** [comment]
Sorry, never touched rust before so this may take a sizable chunk of time to get a dev env set up :)

**Comment by fluential** [comment]
@kvz well you could just run a docker image https://github.com/fluential/docker-tinysearch

**Comment by kvz** [comment]
Unsure how I would change this line

**Comment by kvz** [comment]
Tried this:

- fork your project,
- `cd ~/code`
- `git clone git@github.com:kvz/tinysearch.git`
- `cd tinysearch`
- change line 68 in `storage.rs` from `+ 10` -> `+ 12`
- `cargo run /home/kvz/code/content/_site/tinysearch.json`

Got:

```bash
    Finished dev [unoptimized + debuginfo] target(s) in 0.03s
     Running `target/debug/tinysearch /home/kvz/code/content/_site/tinysearch.json`
Extracting tinysearch WASM engine
Copying index into crate
Compiling WASM module using wasm-pack
Error: Os { code: 2, kind: NotFound, message: "No such file or directory" }

failed to run "wasm-pack" "build" "/tmp/.tmpK6CPU7/engine" "--target" "web" "--release" "--out-dir" "/home/kvz/code/tinysearch"
```

Clearly I have no idea what I'm doing

**Comment by fluential** [comment]
@kvz so if you clone repo, modify the value you can build tinysearch with custom repo url/branch like this
via https://github.com/fluential/docker-tinysearch/blob/master/Dockerfile
```
docker build --build-arg TINY_REPO=<repourl> -t devbuild .
```

**Comment by kvz** [comment]
Tried a few things that way, no dice yet:

```bash
$ cd ~/code
$ git clone git@github.com:kvz/docker-tinysearch.git
$ cd docker-tinysearch/
$ docker build --build-arg TINY_REPO=https://github.com/kvz/tinysearch -t devbuild .
$ docker run -it devbuild:latest

docker run -it devbuild:latest
docker: Error response from daemon: OCI runtime create failed: container_linux.go:348: starting container process caused "exec: \"/usr/local/cargo/bin/tinysearch\": stat /usr/local/cargo/bin/tinysearch: no such file or directory": unknown.

$ docker run -it --volume /site:/home/kvz/code/content/_site devbuild:latest tinysearch /site/tinysearch.json

docker: Error response from daemon: OCI runtime create failed: container_linux.go:348: starting container process caused "exec: \"tinysearch\": executable file not found in $PATH": unknown.

$ docker run -it --volume /site:/home/kvz/code/content/_site devbuild:latest /usr/local/cargo/bin/tinysearch /site/tinysearch.json

docker: Error response from daemon: OCI runtime create failed: container_linux.go:348: starting container process caused "exec: \"/usr/local/cargo/bin/tinysearch\": stat /usr/local/cargo/bin/tinysearch: no such file or directory": unknown.
```

**Comment by kvz** [comment]
Happy to try some more things, but also wondering if making that value `10` configurable would be a more efficient way to obtain datapoints vs trying to make me successfully compile a rust project :)

**Comment by fluential** [comment]
@kvz oops, sorry my bad, please clone the upstream again again and it will build your binary

**Comment by kvz** [comment]
That worked :tada: i tried it with `12` in this case. Still need to integrate and see if i can actually search but there were no errors, and files were produced

**Comment by kvz** [comment]
Hm.. So these files were generated:

```
-rw-r--r--  1 kvz kvz      1 Jun 16 16:11 .gitignore
-rw-r--r--  1 kvz kvz    477 Jun 16 16:11 package.json
-rw-r--r--  1 kvz kvz    279 Jun 16 16:11 tinysearch_engine_bg.d.ts
-rw-r--r--  1 kvz kvz 171772 Jun 16 16:11 tinysearch_engine_bg.wasm
-rw-r--r--  1 kvz kvz    889 Jun 16 16:11 tinysearch_engine.d.ts
-rw-r--r--  1 kvz kvz   4560 Jun 16 16:11 tinysearch_engine.js
```

But not a demo.html so that i could access http://0.0.0.0:8000/demo.html as explained in the README

**Comment by mre** [comment]
> Edit: But when I copy that file manually from https://raw.githubusercontent.com/mre/tinysearch/master/bin/assets/demo.html, searches seem to work 💯

Just tested on my machine and it creates the `demo.html` in the root folder. Odd. 🤔 Maybe others can double-check.

> I wonder if it's a good idea to make that value configurable still. Apparently there's a trade-off for high values, so maybe only people with large indices running into errors should be paying that, while you have a low default that caters to most use cases.

Actually I think the magic number is a bug. This shouldn't be happening and we should fix that. Will keep that issue open so that we can work on it and add a `bug` label.

**Comment by fluential** [comment]
@kvz works for me as well, try this one https://github.com/tinysearch/tinysearch/blob/70c22f70a0377f83becd0c2c3bca42a00fbc2289/Dockerfile
then 
``` docker run $PWD:/tmp tinysearch/cli index.json```

**Comment by Lusitaniae** [comment]
Thanks for the great work @fluential 

Was able to build a search index using magic number 64, end result is 

```
-rw-r--r-- 1 q q 6,7M Jun 11 22:46 index.json
-rw-r--r-- 1 q q 2,7M Jun 19 20:38 tinysearch_engine_bg.wasm
-rw-r--r-- 1 q q 1,1M Jun 19 20:38 tinysearch_engine_bg.wasm.gz
```

This is including all the content, could cut it down if I used the title only.

Probably good enough for me as I'm building media heavy website.

**Comment by mre** [comment]
I want to test what happens with my index size if I set the magic number to 64. If it doesn't blow up the wasm size to much, we can make this the new default while we work on a better solution. I still would like to give XOR filters a try; my hope is that they'll produce even smaller output.
Kudos for getting it working @Lusitaniae and @fluential. If any of you adds tinysearch to their site, please add yourself to the section in the `README.md` to get the word out. Same for you of course, @kvz. 😊

**Comment by lord-re** [comment]
I had the same issue, i had to up the value to 26 to make it work with a 2MB json.

**Comment by mre** [comment]
@lord-re can you try with 64 and comparing the sizes of the output wasm? If it's not much of a difference, I'd make that the default for now.

**Comment by lord-re** [comment]
Going from 26 to 64 the produced files go from 780KB to  868KB.

**Comment by fluential** [comment]
@lord-re can you compare also `gzip -9` sizes?

**Comment by lord-re** [comment]
On which files ? all of them ?

**Comment by lord-re** [comment]
value | uncompressed | gzipped
    26  |             780KB | 456 KB
     64 |             868KB | 472 KB

**Comment by fluential** [comment]
@lord-re just the `*.wasm` thanks!

**Comment by mre** [comment]
Yup, I feel the same. Let's bump it to 64 for now.

**Comment by kvz** [comment]
Would a binary amd64 Linux release with this be in the stars? Would significantly reduce my CI’s complexity :)

**Comment by mre** [comment]
Sure! Will create a new issue for that.

**Comment by mre** [comment]
Actually, I think we can use #15 for that.

**Comment by quangloc1993** [comment]
I resolve the problem NotEnoughSpace,
But how can i custom return value? In my case, my post contain keywords, description, and i want search result contains those values, how can search return with those value?

**Comment by Lusitaniae** [comment]
@quangloc1993 see https://github.com/tinysearch/tinysearch/issues/116

**Comment by quangloc1993** [comment]
@Lusitaniae how can we run it in old version browser, edge and IE, any solution?

**Comment by Lusitaniae** [comment]
@quangloc1993 Better create a new issue regarding that

This project is mostly about generating the search database using rust/wasm, there's only a very basic example of usage.

If you need supporting older browsers, you may need to investigate it yourself...

But let's start with making new issue :+1:

**Comment by quangloc1993** [comment]
@Lusitaniae  can we sort the searrch results, from the best mach to lower?

**Comment by Lusitaniae** [comment]
@quangloc1993 please look at existing issues or create a dedicated issue.

https://github.com/tinysearch/tinysearch/issues/120

**Comment by soyuka** [comment]
I ran into that as well and probably the simple way to handle this is to cut the body, for example with hugo:

```
  {{ range $index , $e := .Site.RegularPages }}{{ if $index }}, {{end}}{{ dict "title" .Title "url" .Permalink "body" ( substr .Plain 0 2000 ) | jsonify }}{{end}}
```

it won't be as accurate ofc. Best would be to find a way to remove the code from the `.Plain` but that looks harder.

**Comment by mre** [comment]
This bug should now be fixed in `master`.
I ported tinysearch over to Xor filters.

* 20-25% smaller wasm size: The test index is 99kB now, 49kB gzipped, 40kB brotli.
(Keep in mind that this was already highly optimized before.)
* Faster initial compilation time thanks to fewer dependencies.
* No more `TINY_MAGIC`. The Xor filter doesn't run into this problem.

Along the way I cleand up the custom serialization code, removed the forked dependencies, which were no longer needed and updated the docs. Oh and I replaced `lazy_static!` with `once_cell!` and `include_bytes!` with the new `include_str`. 
All around good update. 😄 

If anyone wants to give it a shot.
I'll close this to clean up the issue tracker, but feel free to play with the `master` branch and report back here.
